### PR TITLE
CloudProvider: service update event should be triggered when appProtocol in port is changed

### LINKED ISSUE #2434: Health probe is not updated when appProtocol is changed
#### What happened:
Health probe is not updated when appProtocol is changed
#### What you expected to happen:
Health probe should be updated when appProtocol is changed
#### How to reproduce it (as minimally and precisely as possible):
apply svc and change appProtocol. 

#### Environment:

- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration:
- OS (e.g: `cat /etc/os-release`):
- Kernel (e.g. `uname -a`):
- Install tools:
- Network plugin and version (if this is a network-related bug):
- Others:

### PR DESCRIPTION
**Pull request by MartinForReal**
#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:
service update event should be triggered when appProtocol in port is changed

#### Which issue(s) this PR fixes:

Fixes https://github.com/kubernetes-sigs/cloud-provider-azure/issues/2434

#### Does this PR introduce a user-facing change?

```release-note
Fixed a bug where a change in the `appProtocol` for a Service did not trigger a load balancer update.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: staging/src/k8s.io/cloud-provider/controllers/service/controller.go [hunk 1]
```diff
@@ -640,6 +640,11 @@ func portEqualForLB(x, y *v1.ServicePort) bool {
 		return false
 	}
 
+	if (x.AppProtocol == nil) != (y.AppProtocol == nil) ||
```
Comment (andrewsykim) [inline staging/src/k8s.io/cloud-provider/controllers/service/controller.go:643]: This reads slighlty better IMO:

```go
if !reflect.DeepEqual(x.AppProtocol, y.AppProtocol) {
    return false
}
```

**Comment by khenidak** [comment]
/retitle CloudProvider: service update event should be triggered when appProtocol in port is changed

File: staging/src/k8s.io/cloud-provider/controllers/service/controller.go (hunk 1 above)
Comment (MartinForReal) [reply staging/src/k8s.io/cloud-provider/controllers/service/controller.go]: updated

File: staging/src/k8s.io/cloud-provider/controllers/service/controller_test.go [hunk 2]
```diff
@@ -1391,6 +1391,75 @@ func TestNeedsUpdate(t *testing.T) {
...
+						}},
+						Type: v1.ServiceTypeLoadBalancer,
+					},
+				}
+				newSvc = oldSvc.DeepCopy()
+			},
+			expectedNeedsUpdate: false,
+		},
+		{
+			testName: "If appProtocol is different 1",
```
Comment (andrewsykim) [inline staging/src/k8s.io/cloud-provider/controllers/service/controller_test.go:1416]: `If appProtocol is set when previously unset`

File: staging/src/k8s.io/cloud-provider/controllers/service/controller_test.go [hunk 3]
```diff
@@ -1391,6 +1391,75 @@ func TestNeedsUpdate(t *testing.T) {
...
+					},
+				}
+				newSvc = oldSvc.DeepCopy()
+				protocol := "http"
+				newSvc.Spec.Ports[0].AppProtocol = &protocol
+			},
+			expectedNeedsUpdate: true,
+		},
+		{
+			testName: "If appProtocol is different 2",
```
Comment (andrewsykim) [inline staging/src/k8s.io/cloud-provider/controllers/service/controller_test.go:1439]: `If appProtocol is set to a different value`

**Review by andrewsykim — COMMENTED** [review]
Overall LGTM, deferring lgtm to @robscott who worked on adding AppProtocol for Services.

/assign @robscott

File: staging/src/k8s.io/cloud-provider/controllers/service/controller_test.go (hunk 2 above)
Comment (MartinForReal) [reply staging/src/k8s.io/cloud-provider/controllers/service/controller_test.go]: updated.

File: staging/src/k8s.io/cloud-provider/controllers/service/controller_test.go (hunk 3 above)
Comment (MartinForReal) [reply staging/src/k8s.io/cloud-provider/controllers/service/controller_test.go]: updated.

**Comment by MartinForReal** [comment]
<html>
<body>


/test pull-kubernetes-conformance-kind-ga-only-parallel
--


<br class="Apple-interchange-newline">
</body>
</html>

**Comment by robscott** [comment]
Thanks! I can't approve, so will leave that for @andrewsykim. 

/lgtm

**Review by andrewsykim — COMMENTED** [review]
/approve
/lgtm

**Comment by sftim** [comment]
My suggested changelog entry:
```markdown
Fixed a bug where a change in the `appProtocol` for a Service did not trigger a load balancer update.
```

**Pull request merged**